src/hooks/usePortal.ts

```diff
@@ -0,0 +1,303 @@
+    { name: "owner"; type: "address" },
+    { name: "spender"; type: "address" },
+    { name: "value"; type: "uint256" },
+    { name: "nonce"; type: "uint256" },
+    { name: "deadline"; type: "uint256" },
+  ];
+};
+
+type PermitTx = {
+  domain: {
+    name: string;
+    version: string;
+    chainId: number;
+    verifyingContract: `0x${string}`;
+  };
+  types: PermitTypes;
+  value: {
+    owner: `0x${string}`;
+    spender: `0x${string}`;
+    value: bigint;
+    nonce: bigint;
+    deadline: bigint;
+  };
+};
+
+type TransactionToSubmit = {
+  to: `0x${string}`;
+  from: `0x${string}`;
+  data: `0x${string}`;
+  gasLimit: string;
+  value?: bigint;
+};
+
+type PortalQuoteParams = {
+  approveInputToken: UseMutateAsyncFunction;
+  shouldApprove: unknown;
+  gaslessSignature: `0x${string}` | undefined;
+  inputToken: string;
+  inputTokenDecimals: number;
+  inputAmount: string;
+  outputToken: string;
+  sender: string;
+};
+
+export type PortalQuoteResponse = {
+  tx: TransactionToSubmit;
+  context: PortalQuoteContext;
+};
+
+export type PortalQuoteContext = {
+  orderId: string;
+  minOutputAmount: string;
+  minOutputAmountUsd: number;
+  slippageTolerancePercentage: number;
+  gasLimit: string;
+  inputAmount: string;
+  inputAmountUsd: number;
+  inputToken: string;
+  outputToken: string;
+  outputAmount: string;
+  outputAmountUsd: number;
+  partner: string;
+  feeToken: string;
+  feeAmount: string;
+  feeAmountUsd: number;
+  sender: `0x${string}`;
+  recipient: `0x${string}`;
+  target: `0x${string}`;
+  value: string;
+  route: string[];
+  steps: string[];
+};
+
+export type PortalTransactionResult =
+  | { hash: string; signature?: undefined }
+  | { signature: string; hash?: undefined };
+
+const PORTAL_API_BASE_URI = "https://api.portals.fi/v2";
+const PORTALS_CHAIN_ID = {
+  [mainnet.id]: "ethereum",
+  [arbitrum.id]: "arbitrum",
+  [optimism.id]: "optimism",
+  [fantom.id]: "fantom",
+};
+
+const useCheckApproval = (
+  sender: `0x${string}`,
+  inputToken: `0x${string}`,
+  inputAmount: string,
+  inputTokenDecimals: number,
+) => {
+  const chain = useChain();
+
+  return useQuery({
+    queryKey: [
+      "portalCheckApproval",
+      sender,
+      inputToken,
+      inputAmount,
+      inputTokenDecimals,
+    ],
+    queryFn: async () => {
+      const url = new URL(`${PORTAL_API_BASE_URI}/approval`);
+      url.searchParams.append("sender", sender);
+      url.searchParams.append(
+        "inputToken",
+        `${PORTALS_CHAIN_ID[chain.id]}:${inputToken}`,
+      );
+      url.searchParams.append(
+        "inputAmount",
+        `${parseUnits(inputAmount || `0`, inputTokenDecimals)}`,
+      );
+
+      const response = await fetch(url.toString(), { method: "GET" });
+
+      if (!response.ok) {
+        throw new Error(
+          `Error executing call to portal checkApproval: ${response.status} ${response.statusText}`,
+        );
+      }
+
+      const portalRouterApprovalData = await response.json();
```

Comment:
basically, it's a common pattern to assert parsed response here. then it's coupled with the fetch and useQuery or any other abstract fetch function would inherit this assertion